File: app/models/push/Device.java
package models.push;

import java.util.HashSet;
import java.util.Set;


import org.bson.types.ObjectId;

import play.Logger;
import plugin.morphia.MorphiaBootstrapPlugin;

import com.google.code.morphia.Datastore;
import com.google.code.morphia.Key;
import com.google.code.morphia.annotations.Entity;
import com.google.code.morphia.annotations.Id;
import com.google.code.morphia.annotations.PrePersist;
import com.google.code.morphia.annotations.PreSave;
import com.google.code.morphia.annotations.Reference;
import com.mongodb.WriteResult;

@Entity(value = "devices")
public class Device extends BaseEntity {

	static { datastore().ensureIndex(Device.class, "devices_token_type_index", "token, type", true, false); }

	@Id
	public ObjectId id;
	public String token;
	public DeviceType type;
	@Reference
	public Set<Channel> channels = new HashSet<Channel>();

	public Device() {

	}

	public Device(String token, DeviceType type) {
		this.token = token;
		this.type = type;
	}
	
	private static Datastore datastore() { 
		return play.Play.application().plugin(MorphiaBootstrapPlugin.class).getDatastore();
	}

	public static void create(Device device) {
		Key<Device> deviceKey = datastore().save(device);
		device.id = (ObjectId) deviceKey.getId();
		Logger.debug("Created device with key " + deviceKey.getId());
	}

	public static boolean delete(Device device) {
		WriteResult res = datastore().delete(device);
		return res.getN() == 1;
	}

	public static Device findByTokenAndType(String token, DeviceType type) {
		return datastore().createQuery(Device.class).filter("token", scrubToken(token)).filter("type", type.toString()).get();
	}
	
	public static void update(Device device) {
		datastore().save(device);
	}

	@PreSave
	@PrePersist
	public void scrub() {
	    this.token = scrubToken(this.token);
	}
	
	private static String scrubToken(String token) {
	    return token.replaceAll("[\\s<>]", "");
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = super.hashCode();
		result = prime * result + ((id == null) ? 0 : id.hashCode());
		result = prime * result + ((token == null) ? 0 : token.hashCode());
		result = prime * result + ((type == null) ? 0 : type.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (!super.equals(obj))
			return false;
		if (getClass() != obj.getClass())
			return false;
		Device other = (Device) obj;
		if (id == null) {
			if (other.id != null)
				return false;
		} else if (!id.equals(other.id))
			return false;
		if (token == null) {
			if (other.token != null)
				return false;
		} else if (!token.equals(other.token))
			return false;
		if (type != other.type)
			return false;
		return true;
	}

	@Override
	public String toString() {
		return "Device [id=" + id + ", token=" + token + ", type=" + type
				+ ", createdAt=" + createdAt + ", updatedAt=" + updatedAt + "]";
	}
	
}

File: app/push/models/Device.scala
package push.models

import com.mongodb.casbah.Imports._
import com.novus.salat._
import com.novus.salat.annotations._
import com.novus.salat.dao._
import com.novus.salat.global._
import java.util.Date
import play.api.Play.current
import play.api.Logger._
import mongoContext._
import se.radley.plugin.salat._
import DeviceType._
import scala.collection.Set

case class Device(id: ObjectId, token: String, deviceType: DeviceType.Value, channels: Set[ObjectId] = Set[ObjectId](), createdAt: Option[Date] = None, updatedAt: Option[Date] = None)

trait DeviceImpl {
  
  val collection = mongoCollection("Device")
  val dao = new SalatDAO[Device, ObjectId](collection) {}
  
  collection.ensureIndex(MongoDBObject("token" -> 1, "deviceType" -> 1), "token_and_type_index", true)
  
  def findByTokenAndType(token: String, deviceType: DeviceType.Value): Option[Device] = dao.findOne(MongoDBObject("token" -> scrubToken(token), "deviceType" -> deviceType.toString()))
  def create(device: Device) : Option[Device] = {
    try {
      val now = Some(new Date())
      dao.insert(device.copy(token = scrubToken(device.token), createdAt = now, updatedAt = now)).map { id => device.copy(id = id) }
    } catch {
      case e: Exception => error("Failed to create device " + device, e)
      None
    }
  }
  
  def update(device: Device) : Boolean = {
    try {
      info("Attempting to find device " + device)
      dao.findOneById(device.id).map { d:Device =>
        dao.save(device.copy(token = scrubToken(device.token), updatedAt = Some(new Date())))
        info("Saved device!")
        true
      }.getOrElse(false)
    } catch {
      case e: Exception => error("Failed to update device " + device, e)
      false
    }
  }
  
  def delete(device: Device) : Boolean = {
    try {
      dao.findOneById(device.id).map { d:Device =>
        dao.remove(device)
        true
      }.getOrElse(false)
    } catch {
      case e: Exception => error("Failed to remove device " + device + " from database", e)
      false
    }
  }
  
  private def scrubToken(token:String) = token.replaceAll("[\\s<>]", "")
  
}

object Device extends DeviceImpl